### DISCUSSION TITLE
How to get optimal params from optimize method?

### DISCUSSION THREAD

**Discussion by TomaszLakota** [kernc/backtesting.py#232] [Q&A] [answered]
Hi, how do I get optimal params from optimize method? It's not specified in docs. I'd like to know optimal params, so I can use them live or to test them on different dataset (train-test split). I know there's a walkaround using `method = 'skopt'` and `return_optimization = True`, but I want to do exhaustive search.
Edit: I'm using heatmaps to find optimal parameters, but I'd like to automatize it for testing.

**Comment by kernc** [comment, marked as answer]
You can pass [`Backtest.optimize(..., return_heatmap=True)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest.optimize) to get back a series of all tried parameter combinations with their result values. See how to handle it in https://github.com/kernc/backtesting.py/issues/101.
Alternatively, [as the tutorial briefly mentions](https://kernc.github.io/backtesting.py/doc/examples/Quick%20Start%20User%20Guide.html#Optimization), you can access the best parameters as set on the strategy _instance_:
```py
stats = bt.optimize(...)
my_param1 = stats._strategy.my_param1
...
all_params: dict = stats._strategy._params  # This is private API, may change
```

**Reply by TomaszLakota** [reply]
Thanks, I used the method of transforming heatmap into a dataframe to get the best params

**Reply by PilotGFX** [reply]
after returning stats and heatmap from the optimizer, you can use the heatmap to extract the parameters. 

```
best_params = optiheatmap.idxmax()
var1 = best_params[0]
var2 = best_params[1]
var3= best_params[2]
```
You can then either pass these on to update your strategyclass, or just pass them to the backtest as parameters as kernc have explained how to in other threads.
